registry/consul/listener.go

```diff
@@ -0,0 +1,206 @@
+	// upstream.
+	errCh chan error
+
+	// Done field represents whether consul
+	// listener has been closed. When closing
+	// listener, this field will be closed,
+	// and will notify consul watcher to close.
+	done chan struct{}
+
+	// After listener notifies consul watcher
+	// to close, listener will call wg.wait to
+	// make sure that consul watcher is closed
+	// before the listener closes.
+	wg sync.WaitGroup
+}
+
+func newConsulListener(registryUrl common.URL, consumerUrl common.URL) (*consulListener, error) {
+	params := make(map[string]interface{}, 8)
+	params["type"] = "service"
+	params["service"] = consumerUrl.Service()
+	params["tag"] = "dubbo"
+	params["passingonly"] = true
+	plan, err := watch.Parse(params)
+	if err != nil {
+		return nil, err
+	}
+
+	listener := &consulListener{
+		registryUrl: registryUrl,
+		consumerUrl: consumerUrl,
+		plan:        plan,
+		urls:        make([]common.URL, 0, 8),
+		eventCh:     make(chan *registry.ServiceEvent),
+		errCh:       make(chan error),
+		done:        make(chan struct{}),
+	}
+
+	// Set handler to consul watcher, and
+	// make watcher begin to watch service
+	// information change.
+	listener.plan.Handler = listener.handler
+	listener.wg.Add(1)
+	go listener.run()
+	return listener, nil
+}
+
+// Wrap the consul watcher run api. There are three
+// conditions that will finish the run:
+//   - close done
+//   - call plan.Stop
+//   - close eventCh and errCh
+// If run meets first two conditions, it will close
+// gracefully. However, if run meets the last condition,
+// run will close with panic, so use recover to cover
+// this case.
+func (l *consulListener) run() {
+	defer func() {
+		p := recover()
+		if p != nil {
+			logger.Warnf("consul listener finish with panic %v", p)
+		}
+		l.wg.Done()
+	}()
+
+	for {
+		select {
+		case <-l.done:
+			return
+		default:
+			err := l.plan.Run(l.registryUrl.Location)
+			if err != nil {
+				l.errCh <- err
+			}
+		}
+	}
+}
+
+func (l *consulListener) handler(idx uint64, raw interface{}) {
+	var (
+		service *consul.ServiceEntry
+		event   *registry.ServiceEvent
+		url     common.URL
+		ok      bool
+		err     error
+	)
+
+	services, ok := raw.([]*consul.ServiceEntry)
+	if !ok {
+		err = perrors.New("handler get non ServiceEntry type parameter")
+		l.errCh <- err
+		return
+	}
+	newUrls := make([]common.URL, 0, 8)
+	events := make([]*registry.ServiceEvent, 0, 8)
+
+	for _, service = range services {
+		url, err = retrieveURL(service)
+		if err != nil {
+			l.errCh <- err
+			return
+		}
+		newUrls = append(newUrls, url)
+	}
+
+	for _, url = range l.urls {
+		ok = in(url, newUrls)
+		if !ok {
+			event := &registry.ServiceEvent{Action: remoting.EventTypeDel, Service: url}
+			events = append(events, event)
+		}
+	}
+
+	for _, url = range newUrls {
+		ok = in(url, l.urls)
+		if !ok {
+			event := &registry.ServiceEvent{Action: remoting.EventTypeAdd, Service: url}
+			events = append(events, event)
+		}
+	}
+
+	l.urls = newUrls
```

Comment:
need not to add a lock to protect l.urls?

Reply:
Em, the simple reason that why I not use `lock` is that `l.urls` just runs in one goroutine: https://github.com/dubbo-x/dubbo-go/blob/21d7b15a8f7e1deec02086e98f2177b47a1933ba/registry/consul/listener.go#L107

Reply:
ok